### DISCUSSION TITLE
Any plans to upgrade Typescript?

### DISCUSSION THREAD

**Discussion by jwebstar** [eclipse-theia/theia#9628] [Q&A]
Hi, an extension I'm developing depends on a 3rd party package that depends on Typescript 4. Are there any plans to upgrade Theia from ^3.9.2 to a newer Typescript version?

I've been able to get my extension compiling/running with the latest Typescript (4.3.4) by doing the following:

1. theia\packages\core\src\common\messaging\proxy-factory.ts - changing line 239 from:
`"new Promise((resolve, reject)" to: "new Promise<void>((resolve, reject)"`
2. Setting "strictNullChecks" to false in my tsconfig.json (to suppress: "theia/core/src/common/event.ts(333,20): error TS2790: The operand of a 'delete' operator must be optional." - I wasn't sure how to address the underlying cause)

In the meantime could I submit a PR for the 1st point? (or if preferred someone active on the code could commit it)

My extension uses a minimal set of Theia packages, using the latest Typescript with a fork of the main Theia repo does run into more Typescript errors.

**Comment by vince-fugnitto** [comment]
@jwebstar thank you for the discussion, I think in general we should aim to keep the typescript version we use up to date to make use of their improvements and new features. As far as I understand nothing really prevents us from bumping the version like we've done in the [past](https://github.com/eclipse-theia/theia/pull/7798), but we will need to address any build errors that might occur and if both the electron and browser applications work correctly. If you are up for it please feel free to propose a pull-request :+1: 

cc @marcdumais-work @tsmaeder @paul-marechal

**Reply by jwebstar** [reply]
I started an attempt to address the errors with the latest TypeScript version, but quickly realised my TypeScript/Theia toolbelt is a bit too light right now to be able to!

**Comment by paul-marechal** [comment]
As far as I understand, we don't put constraints on what TypeScript version dependents should use. Have you tried setting `skipLibChecks` in your `tsconfig`? I never understood why TypeScript goes and complains about dependencies when you have few control over how declarations are made for it...

**Reply by jwebstar** [reply]
Thanks, I've tried setting skipLibChecks to true but no change. This is my tsconfig:

```jsonc
{
    "compilerOptions": {
        "skipLibCheck": true,
        "declaration": true,
        "declarationMap": true,
        "noImplicitAny": true,
        "noEmitOnError": false,
        "noImplicitThis": true,
        "noUnusedLocals": true,
        "strictNullChecks": false,
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "downlevelIteration": true,
        "resolveJsonModule": true,
        "module": "commonjs",
        "moduleResolution": "node",
        "target": "es5",
        "jsx": "react",
        "lib": [
            "es6",
            "dom"
        ],
        "sourceMap": true,
        "rootDir": "src",
        "outDir": "lib"
    },
    "include": [
        "src"
    ],
    "exclude": [
        "node_modules",
        "src/**/*.test.ts"
    ]
}
```

From what I've just read (e.g. [this](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206634839/comments/115000231764)) I think skipLibCheck and exclude are ignored because I'm importing a file from Theia directly:

```ts
import { ConnectionHandler, JsonRpcConnectionHandler } from '@theia/core';
```

And JsonRpcConnectionHandler is in the file `theia\packages\core\src\common\messaging\proxy-factory.ts` with the code that errors in the newer TypeScript version.

**Comment by tsmaeder** [comment]
>     1. theia\packages\core\src\common\messaging\proxy-factory.ts - changing line 239 from: "new Promise((resolve, reject)" to: "new Promise((resolve, reject)"
> 

I don't see the difference 🤷

**Reply by jwebstar** [reply]
Oops - part of the code wasn't displayed because I didn't put it in a code block - it was meant to be:

`"new Promise((resolve, reject)" to: "new Promise<void>((resolve, reject)"`
